File: sdks/typescript/src/source.ts
import type { SdkPayload } from './runtime/messages.js';
import { ProtocolLimits } from './protocol.js';
import { randomUUID } from 'node:crypto';
import type { PluginContext } from './session.js';

/** A binary source owns its position and returns at most the requested bytes. */
export interface BinarySource {
    read(maxBytes: number): Uint8Array | Promise<Uint8Array>;
    close(): void | Promise<void>;
}
export type SourceHandler = (
    input: any,
    context: PluginContext,
    signal: AbortSignal,
) => BinarySource | Promise<BinarySource>;
export class SourceSession {
    private source?: BinarySource;
    identity?: string;
    async open(
        handler: SourceHandler,
        value: unknown,
        context: PluginContext,
        signal: AbortSignal,
    ): Promise<unknown> {
        const source = await handler(value, context, signal);
        if (typeof source?.read !== 'function' || typeof source?.close !== 'function')
            throw new TypeError('Invalid binary source');
        this.source = context.own(source);
        this.identity = randomUUID();
        return { source: this.identity };
    }
    async read(payload: SdkPayload): Promise<{ data: string; done: boolean }> {
        const size = payload.chunkBytes;
        if (
            typeof size !== 'number' ||
            !Number.isInteger(size) ||
            size < ProtocolLimits.SourceChunkMinimumBytes ||
            size > ProtocolLimits.SourceChunkMaximumBytes
        )
            throw new Error('Source chunk limit');
        const data = await this.source!.read(size);
        if (!(data instanceof Uint8Array) || data.length > size)
            throw new Error('Source must return bounded bytes');
        return {
            data: Buffer.from(data).toString('base64'),
            done: data.length === 0,
        };
    }
    release(): void {
        this.source = undefined;
        this.identity = undefined;
    }
}

File: sdks/python/weaveport_sdk/source.py
"""Exclusive, bounded binary source ownership."""

from .protocol import ProtocolLimits

import base64
import inspect
import uuid


class SourceSession:
    def __init__(self):
        self.source = None
        self.identity = None

    async def open(self, handler, value, context):
        source = handler(value, context)
        if inspect.isawaitable(source):
            source = await source
        if not callable(getattr(source, "read", None)):
            raise TypeError("Source must expose read(size) and close() or aclose()")
        context.own(source)
        self.source, self.identity = source, uuid.uuid4().hex
        return dict(source=self.identity)

    async def read(self, payload):
        size = payload.get("chunkBytes")
        if (
            type(size) is not int
            or not ProtocolLimits.SourceChunkMinimumBytes
            <= size
            <= ProtocolLimits.SourceChunkMaximumBytes
        ):
            raise ValueError("Source chunk limit")
        data = self.source.read(size)
        if inspect.isawaitable(data):
            data = await data
        if not isinstance(data, (bytes, bytearray)) or len(data) > size:
            raise ValueError("Source must return bounded bytes")
        return dict(data=base64.b64encode(data).decode("ascii"), done=len(data) == 0)

    def release(self):
        self.source = self.identity = None
